### PR TITLE
Extensions — subject alternative name

### PR DESCRIPTION
**Pull request by fengb**
Many commercial certificates rely on SAN to support multiple domains. Example for wikipedia:
- Common Name — *.wikipedia.org
- SAN DNS Name — *.m.wikipedia.org
- SAN DNS Name — *.m.mediawiki.org
- _etc_

This PR adds plumbing to parse extensions and enables iterable SAN.

### CODE REVIEW

File: src/main.zig [hunk 1]
```diff
@@ -609,10 +669,48 @@ const ServerCertificate = struct {
     bytes: []const u8,
     dn: []const u8,
     common_name: []const u8,
+    raw_subject_alternative_name: []const u8,
     public_key: x509.PublicKey,
     signature: asn1.BitString,
     signature_algorithm: SignatureAlgorithm,
     is_ca: bool,
+
+    const GeneralName = enum(u5) {
```
Comment (fengb) [inline src/main.zig:678]: This is a lousy name; I adopted it from the spec.

File: src/main.zig [hunk 2]
```diff
@@ -609,10 +669,48 @@ const ServerCertificate = struct {
     bytes: []const u8,
     dn: []const u8,
     common_name: []const u8,
+    raw_subject_alternative_name: []const u8,
```
Comment (fengb) [inline src/main.zig:672]: I wanted to pre-parse this, but unfortunately that requires adding a new allocation (or a large enough buffer) so this adopts an iterator approach instead.

File: src/main.zig [hunk 3]
```diff
@@ -353,6 +353,64 @@ fn add_cert_public_key(state: *VerifierCaptureState, _: u8, length: usize, reade
...
+    // TODO: add a preflight check to mandate a specific tag
+    const object_id = try asn1.der.parse_value(state.allocator, reader);
+    defer object_id.deinit(state.allocator);
+    if (object_id != .object_identifier) return error.DoesNotMatchSchema;
+    if (object_id.object_identifier.len != 4)
+        return;
+
+    const data = object_id.object_identifier.data;
+    // Prefix == id-ce
+    if (data[0] != 2 or data[1] != 5 or data[2] != 29)
```
Comment (fengb) [inline src/main.zig:381]: These magic numbers are copied from x509. We should probably define them somewhere >_>

File: src/main.zig (hunk 2 above)
Comment (fengb) [reply src/main.zig:672]: Also this name sucks

File: src/main.zig (hunk 1 above)
Comment (alexnask) [reply src/main.zig:678]: The field names should be in snake_case

**Review by alexnask — APPROVED** [review]
Very cool! Thanks

File: src/main.zig (hunk 3 above)
Comment (alexnask) [reply src/main.zig:381]: Yes, I will go over the source code once I add client certificates and pull a bunch of the object identifiers we check against out into a new file.

**Pull request merged**